### PR TITLE
Added voice analyzer & Entertainment monitor supporting audible emotes

### LINKED ISSUE #2815: Voice Analyzers and Emotes
Voice analyzers are unable to detect emotes and register them as something said/stated as a potential trigger for them---thereby limiting the scope of their effectiveness.

### PR DESCRIPTION
**Pull request by Tastyfish**
There wasn't actually a standard way to do this before, so...
* I added a new proc `/atom/proc/hear_message(mob, message)` that is used like `hear_talk()`, but is for hearing audible emotes.

##### This is noticably used by...
* The Voice Analyzer
 * Fixes #2815 
* Video Camera -> Entertainment Monitor
* Tape Recorder
* AI Hologram thing
* More objects could probably use this, but I'm going to consider that outside the scope of this PR unless there's a major omission I'm not thinking of.

Also made megaphone slightly less broken.

### CODE REVIEW

**Comment by Fox-McCloud** [comment]
Oh, that's awesome; good thinking for the video-camera; that would make perfect sense, too--so would the tape recorder.

Thanks a million!

**Comment by Ar3nn** [comment]
The only video camera I know of is on ERT helmets. What's the point of those?

**Comment by Fox-McCloud** [comment]
The librarian has a camera.

**Comment by ZomgPonies** [comment]
Journalist alt-job for Librarian...has an actual tv camera he carries around.

**Comment by Ar3nn** [comment]
I mean, specifically, whats the point of the ERT helmet cams?

**Comment by ZomgPonies** [comment]
Ah, ERT leader or admin Centcom officer can monitor their progress remotely.

**Comment by Ar3nn** [comment]
Oh, so its like the Warden using the security cameras, but on people. Gotcha! I'll leave now, I'm obstructing actual work being done, probably.

File: code/modules/paperwork/photography.dm [hunk 1]
```diff
@@ -523,14 +522,19 @@ var/list/SpookyGhosts = list("ghost","shade","shade2","ghost-narsie","horror","s
...
 		user << "This video camera can send live feeds to the entertainment network. It's [camera ? "" : "in"]active."
 
-
 /obj/item/device/videocam/hear_talk(mob/M as mob, msg)
 	if (camera && on)
 		if(get_dist(src, M) <= canhear_range)
 			talk_into(M, msg)
-		for(var/mob/living/carbon/human/H in watcherslist)
-			H.show_message(text("\blue (Newscaster) [] says, '[]'",M,msg), 1)
+		for(var/obj/machinery/computer/security/telescreen/T)
```
Comment (tigercat2000) [inline code/modules/paperwork/photography.dm]: This is missing the `in` operator.

File: code/modules/paperwork/photography.dm [hunk 2]
```diff
@@ -523,14 +522,19 @@ var/list/SpookyGhosts = list("ghost","shade","shade2","ghost-narsie","horror","s
...
+			if(T.current == camera)
+				T.audible_message("<span class='game radio'><span class='name'>(Newscaster) [M]</span> says, '[msg]'", hearing_distance = 2)
 
+/obj/item/device/videocam/hear_message(mob/M as mob, msg)
+	if (camera && on)
+		for(var/obj/machinery/computer/security/telescreen/T)
```
Comment (tigercat2000) [inline code/modules/paperwork/photography.dm]: as is this

**Comment by Plotron** [comment]
How about intercoms/bounced radios?
And how do megaphones work exactly? AFAIR AIs using holopads are unable to understand/hear the megaphone, but that'd have to be checked.

Well yeah, it's outside the scope of this PR, sure.

**Comment by Fox-McCloud** [comment]
For Intecomms/bounced radios, you have the ! thing.

File: code/modules/paperwork/photography.dm (hunk 1 above)
Comment (Tastyfish) [inline code/modules/paperwork/photography.dm]: If you omit `in ...`, it goes through all of the atoms of that type in the world. This is intentional.

File: code/modules/paperwork/photography.dm (hunk 2 above)
Comment (Tastyfish) [inline code/modules/paperwork/photography.dm]: See above

**Comment by Tastyfish** [comment]
Oh yeah! I actually meant to include tape recorders originally, but forgot about them. Will add those. Radios are really complex, I was somewhat intentionally avoiding that, but I might look at it again. The megaphones not being hearable by the holopad sounds like a straight-out bug, so I'll look at that.

File: code/modules/paperwork/photography.dm (hunk 1 above)
Comment (tigercat2000) [inline code/modules/paperwork/photography.dm]: if you want it to loop through the world, use `in world`, because that's much more readable

**Comment by Fox-McCloud** [comment]
@Tastyfish 

I wouldn't worry about radios, to be honest; we have the function where if you include the exclamation, then it prints the emote over radio---in my opinion, that's good enough, especially given how ridiculously complex radios and saycode are.

That's my opinion anyway--dunno for 100% sure.

**Comment by Plotron** [comment]
Fox, but wouldn't it be cool to have your screams heard on the intercom/bounced radio even if they were involuntary (that is - without ! )?
Sure, feasiblity of coding this is debatable. Probably not worth the effort.

**Comment by Fox-McCloud** [comment]
@Plotron 

Not really, to be honest; that would make murdering someone even more ridiculously hard than it currently is.

**Comment by Tastyfish** [comment]
@Fox-McCloud Oh welp, I already made open broadcasting radios pick up emotes before you responded, I can cut it out again if it's a balance issue.

File: code/modules/paperwork/photography.dm (hunk 1 above)
Comment (Tastyfish) [inline code/modules/paperwork/photography.dm]: OK

**Comment by Fox-McCloud** [comment]
I do believe there are certain instances where that can occur---not to mention that if you try to scream and can't, I believe it returns "makes a loud/strong noise", which I think is still audible and picked up---so that's why I'm primarily concerned about radios.

I don't really have an issue with megaphones or holograms--I think that's a great addition; it's just radios that I find problematic.

**Comment by Tastyfish** [comment]
I will cut out them picking emotes up then.

**Comment by ZomgPonies** [comment]
If you're looping through telecoms stuff. Don't loop through in world which will check every atom. loop instead through in machines to only look through the machinery list. Better performance. "in world" loops are the death of the server.

**Comment by Tastyfish** [comment]
I assumed that it optimized type filtered loops, but I can change to machines.

**Comment by Tastyfish** [comment]
Not sure why I assumed anything in the BYOND engine would be optimized.

File: code/game/machinery/hologram.dm [hunk 3]
```diff
@@ -123,6 +123,13 @@ For the other part of the code, check silicon say.dm. Particularly robot talk.*/
 			rendered = "<i><span class='game say'>Holopad received, <span class='name'>[name_used]</span> [verb], <span class='message'>\"[text]\"</span></span></i>"
 		master.show_message(rendered, 2)
 	return
+	
+/obj/machinery/hologram/holopad/hear_message(mob/living/M, text)
+	if(M&&hologram&&master)//Master is mostly a safety in case lag hits or something.
+		var/name_used = M.GetVoice()
+		var/rendered = "<i><span class='game say'>Holopad received, <span class='name'>[name_used]</span> [text]</span></i>"
```
Comment (TheDZD) [inline code/game/machinery/hologram.dm]: I'm a bit unsure of how spans will react to being within spans. Have you checked to make sure this doesn't do anything odd?

Comment (Tastyfish) [inline code/game/machinery/hologram.dm]: Those classes of spans are specifically meant to be used in that exact way.

**Pull request merged**